dhis2/dhis2-core
# Data Integrity checks (2.37)

### Summary
This is a backport of data integrity domain code to 2.37 from master "as is".
That means majority of files were copied using `git checkout master <file>`.

As data integrity builds upon newst scheduling I also had to backport the scheduling domain in a similar way.

Both data integrity and scheduling also required new caches so the cache domain needed selective updates.

All in all little manual adjustment was needed after the initial copying. 
Most of it due to 2 key differences:
* Java 11 vs Java 8 (`List.of` => `Arrays.asList` and `toUnmodifiableList()` => `toList()`)
* junit5 vs junit4

### Automatic Testing
Existing tests were migrated back to junit4.
2 test scenarios needed to be `@Ignore`d but the issue was transaction boundaries in test setup.

### Manual Testing
* create and run analytics table job manually
* create and run resource table job manually
* create and run data integrity job manually and check `/api/scheduling/running/DATA_INTEGRITY` and `/api/scheduling/completed/DATA_INTEGRITY`
* run `/dhis-web-data-administration/index.html#/analytics` and check steps are shown and finish
* run `/dhis-web-data-administration/index.html#/resourceTables` and check steps are shown and finish
* trigger check: `POST` `/api/dataIntegrity/summary?checks=categories_unique_category_combo` and check results `GET` `/api/dataIntegrity/summary?checks=categories_unique_category_combo`
* trigger check: `POST` `/api/dataIntegrity/details?checks=categories_unique_category_combo` and check results `GET` `/api/dataIntegrity/details?checks=categories_unique_category_combo`
